demos/demo.cpp

```diff
@@ -0,0 +1,171 @@
+/**
+ * @file demo.cpp
+ * @author Vincent Berenz
+ * license License BSD-3-Clause
+ * @copyright Copyright (c) 2019, Max Planck Gesellschaft.
+ * 
+ * @brief Minimal demo of robot driver, backend and frontend
+ */
+
+#include "robot_interfaces/robot_driver.hpp"
+#include "robot_interfaces/robot_backend.hpp"
+#include "robot_interfaces/robot_frontend.hpp"
+
+#include <memory>
+
+/**
+ * \example demo.cpp
+ * This demo shows robot_interfaces of a 
+ * dummy "2dof" robot, in which a dof "position"
+ * is represented by an integer
+ */
+
+
+
+// Actions to be performed by robot, will be received by Driver
+// An action simply encapsulate two desired position value,
+// one for each dof
+class Action
+{
+  
+public:
+  
+  int values[2];
+
+  void print(bool backline)
+  {
+    std::cout << "action: "
+	      << values[0] << " "
+	      << values[1] << " ";
+    if(backline)
+      std::cout << "\n";
+  }
+  
+};
+
+// Read from the robot by Driver
+// An observation is the current position
+// for each dof
+class Observation
+{
+  
+public:
+  
+  int values[2];
+
+  void print(bool backline)
+  {
+    std::cout << "observation: "
+	      << values[0] << " "
+	      << values[1] << " ";
+    if(backline)
+      std::cout << "\n";
+  }
+  
+};
+
+
+// Send command to the robot and read observation from the robot
+// The dof positions simply becomes the ones set by the latest action,
+// capped between a min and a max value (0 and 1000)
+class Driver : public robot_interfaces::RobotDriver< Action,Observation>
+{
+  
+public:
+
+  Driver(){}
+
+  // at init dof are at min value
+  void initialize()
+  {
+    values_[0] = Driver::MIN;
+    values_[1] = Driver::MIN;
+  }
+
+  // just clip desired values
+  // between 0 and 1000
+  Action apply_action(const Action &desired_action)
+  {
+    Action applied;
+    for(unsigned int i=0;i<2;i++)
+      {
+	if(desired_action.values[i]>Driver::MAX)
+	  {
+	    applied.values[i]=Driver::MAX;
+	  }
+	else if(desired_action.values[i]<Driver::MIN)
+	  {
+	    applied.values[i]=Driver::MIN;
+	  }
+	else
+	  {
+	    applied.values[i]=desired_action.values[i];
+	  }
```

Comment:
you could use the [clamp](https://github.com/open-dynamic-robot-initiative/mpi_cpp_tools/blob/master/include/mpi_cpp_tools/math.hpp#L17) function from mpi_cpp_tools to do all of the above in one line :)

Reply:
I was thinking about this as well. It would make it more readable but at the cost of adding a dependency and being less self-contained. So I am unsure which way is better.

Reply:
I do not like the idea of adding a dependency for such a trivial matter

Reply:
we already depend on this package, do you mean including the header?

Reply:
maybe, but I do not know this package enough to know what dependency for which reason. I just know that using an external package to perform a trivial clampling is not healthy. I also believe this a little too fine grained of a code review.